crates/ty_python_semantic/resources/mdtest/public_types.md

```diff
@@ -0,0 +1,421 @@
+
+def outer(x: C | None):
+    x = x or C()
+
+    reveal_type(x)  # revealed: C
+
+    def inner() -> None:
+        # TODO: this should ideally be `C`
+        reveal_type(x)  # revealed: C | None
+    inner()
+```
+
+### Assignments to nonlocal variables
+
+Writes to the outer-scope variable are currently not detected:
+
+```py
+def outer() -> None:
+    x = None
+
+    def set_x() -> None:
+        nonlocal x
+        x = 1
+    set_x()
+
+    def inner() -> None:
+        # TODO: this should ideally be `Unknown | None | Literal[1]`.
+        reveal_type(x)  # revealed: Unknown | None
+    inner()
+```
+
+## Handling of overloads
+
+### With implementation
+
+Overloads need special treatment, because here, we do not want to consider *all* possible
+definitions of `f`. This would otherwise result in a union of all three definitions of `f`:
+
+```py
+from typing import overload
+
+@overload
+def f(x: int) -> int: ...
+@overload
+def f(x: str) -> str: ...
+def f(x: int | str) -> int | str:
+    raise NotImplementedError
+
+reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+
+def _():
+    reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+```
+
+This also works if there are conflicting declarations:
+
+```py
+def flag() -> bool:
+    return True
+
+if flag():
+    @overload
+    def g(x: int) -> int: ...
+    @overload
+    def g(x: str) -> str: ...
+    def g(x: int | str) -> int | str:
+        return x
+
+else:
+    g: str = ""
+
+def _():
+    reveal_type(g)  # revealed: (Overload[(x: int) -> int, (x: str) -> str]) | str
+
+# error: [conflicting-declarations]
+g = "test"
+```
+
+### Without an implementation
+
+If there is no implementation, we only consider the last overload definition.
+
+```pyi
+from typing import overload
+
+@overload
+def f(x: int) -> int: ...
+@overload
+def f(x: str) -> str: ...
+
+reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+
+def _():
+    reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+```
+
+This also works if there are conflicting declarations:
+
+```pyi
+def flag() -> bool:
+    return True
+
+if flag():
+    @overload
+    def g(x: int) -> int: ...
+    @overload
+    def g(x: str) -> str: ...
+else:
+    g: str
+
+def _():
+    reveal_type(g)  # revealed: (Overload[(x: int) -> int, (x: str) -> str]) | str
+```
+
+### Conditional overloads
+
+```py
+from typing import overload
+
+def flag() -> bool:
+    return True
+
+@overload
+def f(x: int) -> int: ...
+
+if flag():
+    @overload
+    def f(x: str) -> str: ...
+
+else:
+    @overload
+    def f(x: bytes) -> bytes: ...
+
+reveal_type(f)  # revealed: (Overload[(x: int) -> int, (x: str) -> str]) | (Overload[(x: int) -> int, (x: bytes) -> bytes])
+
+def _():
+    # TODO: ideally, this should be the same union type as above.
+    # revealed: (def f(x: int) -> int) | (Overload[(x: int) -> int, (x: str) -> str]) | (Overload[(x: int) -> int, (x: bytes) -> bytes])
```

Comment:
hmm, this one's a bit concerning

Reply:
it still seems like it would be highly confusing if you came across it, but I do agree that it should be rare.